File: packages/game-engine/src/evaluator.ts
import type { CardCode } from "@parker/protocol";

import { parseCard, type Card } from "./cards.js";

export interface HandScore {
  category: number;
  rankValues: number[];
  bestFive: CardCode[];
  label: string;
}

const HAND_LABELS = [
  "high-card",
  "pair",
  "two-pair",
  "three-of-a-kind",
  "straight",
  "flush",
  "full-house",
  "four-of-a-kind",
  "straight-flush",
] as const;

function compareRankArrays(left: number[], right: number[]): number {
  const maxLength = Math.max(left.length, right.length);
  for (let index = 0; index < maxLength; index += 1) {
    const leftValue = left[index] ?? 0;
    const rightValue = right[index] ?? 0;
    if (leftValue > rightValue) {
      return 1;
    }
    if (leftValue < rightValue) {
      return -1;
    }
  }
  return 0;
}

function sortCardsDesc(cards: Card[]): Card[] {
  return [...cards].sort((left, right) => right.rankValue - left.rankValue);
}

function getStraightHigh(sortedDistinctRanksDesc: number[]): number | null {
  const ranks = [...sortedDistinctRanksDesc];
  if (ranks[0] === 14) {
    ranks.push(1);
  }

  let run = 1;
  let bestHigh: number | null = null;

  for (let index = 0; index < ranks.length - 1; index += 1) {
    if (ranks[index] === ranks[index + 1]! + 1) {
      run += 1;
      if (run >= 5) {
        bestHigh = ranks[index - 3]!;
      }
    } else {
      run = 1;
    }
  }

  return bestHigh;
}

function compareFiveCardHands(left: HandScore, right: HandScore): number {
  if (left.category > right.category) {
    return 1;
  }
  if (left.category < right.category) {
    return -1;
  }
  return compareRankArrays(left.rankValues, right.rankValues);
}

function evaluateFiveCardHand(cards: Card[]): HandScore {
  const sorted = sortCardsDesc(cards);
  const suits = new Set(sorted.map((card) => card.suit));
  const isFlush = suits.size === 1;
  const counts = new Map<number, number>();

  for (const card of sorted) {
    counts.set(card.rankValue, (counts.get(card.rankValue) ?? 0) + 1);
  }

  const entries = [...counts.entries()].sort((left, right) => {
    if (right[1] !== left[1]) {
      return right[1] - left[1];
    }
    return right[0] - left[0];
  });

  const distinctRanksDesc = [...counts.keys()].sort((left, right) => right - left);
  const straightHigh = getStraightHigh(distinctRanksDesc);

  if (isFlush && straightHigh) {
    return {
      category: 8,
      rankValues: [straightHigh],
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[8],
    };
  }

  if (entries[0]?.[1] === 4) {
    const quad = entries[0][0];
    const kicker = entries[1]![0];
    return {
      category: 7,
      rankValues: [quad, kicker],
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[7],
    };
  }

  if (entries[0]?.[1] === 3 && entries[1]?.[1] === 2) {
    return {
      category: 6,
      rankValues: [entries[0][0], entries[1][0]],
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[6],
    };
  }

  if (isFlush) {
    return {
      category: 5,
      rankValues: sorted.map((card) => card.rankValue),
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[5],
    };
  }

  if (straightHigh) {
    return {
      category: 4,
      rankValues: [straightHigh],
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[4],
    };
  }

  if (entries[0]?.[1] === 3) {
    const kickers = entries.slice(1).map(([rank]) => rank).sort((left, right) => right - left);
    return {
      category: 3,
      rankValues: [entries[0][0], ...kickers],
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[3],
    };
  }

  if (entries[0]?.[1] === 2 && entries[1]?.[1] === 2) {
    const pairs = [entries[0][0], entries[1][0]].sort((left, right) => right - left);
    const kicker = entries[2]![0];
    return {
      category: 2,
      rankValues: [...pairs, kicker],
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[2],
    };
  }

  if (entries[0]?.[1] === 2) {
    const kickers = entries.slice(1).map(([rank]) => rank).sort((left, right) => right - left);
    return {
      category: 1,
      rankValues: [entries[0][0], ...kickers],
      bestFive: sorted.map((card) => card.code),
      label: HAND_LABELS[1],
    };
  }

  return {
    category: 0,
    rankValues: sorted.map((card) => card.rankValue),
    bestFive: sorted.map((card) => card.code),
    label: HAND_LABELS[0],
  };
}

function combinations<T>(items: T[], size: number): T[][] {
  if (size === 0) {
    return [[]];
  }
  if (items.length < size || items.length === 0) {
    return [];
  }

  const head = items[0]!;
  const tail = items.slice(1);
  return [
    ...combinations(tail, size - 1).map((combination) => [head, ...combination]),
    ...combinations(tail, size),
  ];
}

export function scoreSevenCardHand(cardCodes: CardCode[]): HandScore {
  if (cardCodes.length !== 7) {
    throw new Error(`expected 7 cards, received ${cardCodes.length}`);
  }

  let best: HandScore | null = null;

  for (const candidate of combinations(cardCodes.map(parseCard), 5)) {
    const score = evaluateFiveCardHand(candidate);
    if (!best || compareFiveCardHands(score, best) > 0) {
      best = score;
    }
  }

  if (!best) {
    throw new Error("could not score seven card hand");
  }

  return best;
}

export function compareScoredHands(left: HandScore, right: HandScore): number {
  return compareFiveCardHands(left, right);
}

File: internal/game/evaluator.go
package game

import (
	"sort"
)

type HandScore struct {
	Category   int
	RankValues []int
	BestFive   []CardCode
	Label      string
}

var handLabels = []string{
	"high-card",
	"pair",
	"two-pair",
	"three-of-a-kind",
	"straight",
	"flush",
	"full-house",
	"four-of-a-kind",
	"straight-flush",
}

func compareRankArrays(left, right []int) int {
	maxLength := len(left)
	if len(right) > maxLength {
		maxLength = len(right)
	}
	for index := 0; index < maxLength; index++ {
		leftValue := 0
		if index < len(left) {
			leftValue = left[index]
		}
		rightValue := 0
		if index < len(right) {
			rightValue = right[index]
		}
		if leftValue > rightValue {
			return 1
		}
		if leftValue < rightValue {
			return -1
		}
	}
	return 0
}

func sortCardsDesc(cards []Card) []Card {
	clone := append([]Card(nil), cards...)
	sort.SliceStable(clone, func(left, right int) bool {
		return clone[left].RankValue > clone[right].RankValue
	})
	return clone
}

func getStraightHigh(sortedDistinctRanksDesc []int) *int {
	ranksCopy := append([]int(nil), sortedDistinctRanksDesc...)
	if len(ranksCopy) > 0 && ranksCopy[0] == 14 {
		ranksCopy = append(ranksCopy, 1)
	}

	run := 1
	var bestHigh *int
	for index := 0; index < len(ranksCopy)-1; index++ {
		if ranksCopy[index] == ranksCopy[index+1]+1 {
			run++
			if run >= 5 {
				value := ranksCopy[index-3]
				bestHigh = &value
			}
		} else {
			run = 1
		}
	}
	return bestHigh
}

func compareFiveCardHands(left, right HandScore) int {
	if left.Category > right.Category {
		return 1
	}
	if left.Category < right.Category {
		return -1
	}
	return compareRankArrays(left.RankValues, right.RankValues)
}

func evaluateFiveCardHand(cards []Card) HandScore {
	sorted := sortCardsDesc(cards)
	suits := map[string]struct{}{}
	for _, card := range sorted {
		suits[card.Suit] = struct{}{}
	}
	isFlush := len(suits) == 1

	counts := map[int]int{}
	for _, card := range sorted {
		counts[card.RankValue]++
	}

	type entry struct {
		rank  int
		count int
	}

	entries := make([]entry, 0, len(counts))
	for rank, count := range counts {
		entries = append(entries, entry{rank: rank, count: count})
	}
	sort.Slice(entries, func(left, right int) bool {
		if entries[left].count != entries[right].count {
			return entries[left].count > entries[right].count
		}
		return entries[left].rank > entries[right].rank
	})

	distinctRanksDesc := make([]int, 0, len(counts))
	for rank := range counts {
		distinctRanksDesc = append(distinctRanksDesc, rank)
	}
	sort.Slice(distinctRanksDesc, func(left, right int) bool {
		return distinctRanksDesc[left] > distinctRanksDesc[right]
	})

	straightHigh := getStraightHigh(distinctRanksDesc)

	bestFive := make([]CardCode, len(sorted))
	for index, card := range sorted {
		bestFive[index] = card.Code
	}

	if isFlush && straightHigh != nil {
		return HandScore{
			Category:   8,
			RankValues: []int{*straightHigh},
			BestFive:   bestFive,
			Label:      handLabels[8],
		}
	}

	if len(entries) > 0 && entries[0].count == 4 {
		quad := entries[0].rank
		kicker := entries[1].rank
		return HandScore{
			Category:   7,
			RankValues: []int{quad, kicker},
			BestFive:   bestFive,
			Label:      handLabels[7],
		}
	}

	if len(entries) > 1 && entries[0].count == 3 && entries[1].count == 2 {
		return HandScore{
			Category:   6,
			RankValues: []int{entries[0].rank, entries[1].rank},
			BestFive:   bestFive,
			Label:      handLabels[6],
		}
	}

	if isFlush {
		rankValues := make([]int, 0, len(sorted))
		for _, card := range sorted {
			rankValues = append(rankValues, card.RankValue)
		}
		return HandScore{
			Category:   5,
			RankValues: rankValues,
			BestFive:   bestFive,
			Label:      handLabels[5],
		}
	}

	if straightHigh != nil {
		return HandScore{
			Category:   4,
			RankValues: []int{*straightHigh},
			BestFive:   bestFive,
			Label:      handLabels[4],
		}
	}

	if len(entries) > 0 && entries[0].count == 3 {
		kickers := make([]int, 0, len(entries)-1)
		for _, candidate := range entries[1:] {
			kickers = append(kickers, candidate.rank)
		}
		sort.Slice(kickers, func(left, right int) bool {
			return kickers[left] > kickers[right]
		})
		rankValues := append([]int{entries[0].rank}, kickers...)
		return HandScore{
			Category:   3,
			RankValues: rankValues,
			BestFive:   bestFive,
			Label:      handLabels[3],
		}
	}

	if len(entries) > 1 && entries[0].count == 2 && entries[1].count == 2 {
		pairs := []int{entries[0].rank, entries[1].rank}
		sort.Slice(pairs, func(left, right int) bool {
			return pairs[left] > pairs[right]
		})
		kicker := entries[2].rank
		return HandScore{
			Category:   2,
			RankValues: []int{pairs[0], pairs[1], kicker},
			BestFive:   bestFive,
			Label:      handLabels[2],
		}
	}

	if len(entries) > 0 && entries[0].count == 2 {
		kickers := make([]int, 0, len(entries)-1)
		for _, candidate := range entries[1:] {
			kickers = append(kickers, candidate.rank)
		}
		sort.Slice(kickers, func(left, right int) bool {
			return kickers[left] > kickers[right]
		})
		rankValues := append([]int{entries[0].rank}, kickers...)
		return HandScore{
			Category:   1,
			RankValues: rankValues,
			BestFive:   bestFive,
			Label:      handLabels[1],
		}
	}

	rankValues := make([]int, 0, len(sorted))
	for _, card := range sorted {
		rankValues = append(rankValues, card.RankValue)
	}
	return HandScore{
		Category:   0,
		RankValues: rankValues,
		BestFive:   bestFive,
		Label:      handLabels[0],
	}
}

func combinations[T any](items []T, size int) [][]T {
	if size == 0 {
		return [][]T{{}}
	}
	if len(items) < size || len(items) == 0 {
		return nil
	}

	head := items[0]
	tail := items[1:]

	withHead := combinations(tail, size-1)
	for index := range withHead {
		withHead[index] = append([]T{head}, withHead[index]...)
	}

	return append(withHead, combinations(tail, size)...)
}

func ScoreSevenCardHand(cardCodes []CardCode) (HandScore, error) {
	if len(cardCodes) != 7 {
		return HandScore{}, fmtErrorf("expected 7 cards, received %d", len(cardCodes))
	}

	var best *HandScore
	for _, candidate := range combinations(cardCodes, 5) {
		cards := make([]Card, 0, len(candidate))
		for _, code := range candidate {
			cards = append(cards, ParseCard(code))
		}
		score := evaluateFiveCardHand(cards)
		if best == nil || compareFiveCardHands(score, *best) > 0 {
			copyScore := score
			best = &copyScore
		}
	}

	if best == nil {
		return HandScore{}, fmtErrorf("could not score seven card hand")
	}

	return *best, nil
}

func CompareScoredHands(left, right HandScore) int {
	return compareFiveCardHands(left, right)
}
